include/rocksdb/utilities/object_registry.h

```diff
@@ -76,29 +131,75 @@ class ObjectLibrary {
     AddEntry(T::Type(), entry);
     return factory;
   }
+
+  // Invokes the registrar function with the supplied arg for this library.
+  void Register(const RegistrarFunc& registrar, const std::string& arg) {
+    registrar(*this, arg);
+  }
+
   // Returns the default ObjectLibrary
   static std::shared_ptr<ObjectLibrary>& Default();
 
+ protected:
+  friend class ObjectRegistry;
+  // ** FactoryFunctions for this loader, organized by type
+  std::unordered_map<std::string, std::vector<std::unique_ptr<Entry>>> entries_;
+  // Returns the
+  std::string AsPrintableOptions(const std::string& name) const;
+
  private:
   // Adds the input entry to the list for the given type
   void AddEntry(const std::string& type, std::unique_ptr<Entry>& entry);
-
-  // ** FactoryFunctions for this loader, organized by type
-  std::unordered_map<std::string, std::vector<std::unique_ptr<Entry>>> entries_;
 };
 
 // The ObjectRegistry is used to register objects that can be created by a
 // name/pattern at run-time where the specific implementation of the object may
 // not be known in advance.
-class ObjectRegistry {
+class ObjectRegistry : public Configurable {
  public:
   static std::shared_ptr<ObjectRegistry> NewInstance();
 
-  ObjectRegistry();
+  // Makes a copy of this registry.
+  virtual std::shared_ptr<ObjectRegistry> Clone() const = 0;
 
-  void AddLibrary(const std::shared_ptr<ObjectLibrary>& library) {
-    libraries_.emplace_back(library);
-  }
+  // Creates a new local library, registering the factories in registrar
+  // @param registrar   The registration function to invoke
+  // @param method      The name of the registration function
+  // @param arg         Argument to supply to the registration function.
+  virtual void AddLocalLibrary(const RegistrarFunc& registrar,
```

Comment:
I'm little confused by the two "AddLocalLibrary" and two "AddDynamicLibrary". Can you explain more?

Reply:
For local, the one signature takes in a name (the method argument) of a registrar function and attempts to find that function in the executable using the Env DynamicLibrary methods.  If the function was not found, an error is returned.  

The other method takes in the RegistrarFunction (a std::function) and invokes it to register the types in the library.  The first method will end up calling this one if the function is found.

Comparable signatures exist for the dynamic libraries, using a DynamicLibrary to locate the registrar function and the Env to locate the dynamic library.

The "string-based" signature must be used in order to recover/restore the ObjectRegistry from a string (or options file).  The "registrar" signature is what does the actual registration, and is useful for programs who wish to register their types at "compile-time".

Reply:
Thanks for the explanation. Maybe you can add those to the comments of "AddLocalLibrary" and "AddDynamicLibrary"